ImGuiIntegration: add support for dynamic textures

pezcode opened the pull request:
Adds support for dynamic texture creation and updating to the ImGui backend. This got added to ImGui in version 1.92 and mainly allows dynamic font atlasses for lower memory and font/DPI scaling without having to rebuild the whole atlas.

This sort of (but not really, see below) obsoletes https://github.com/mosra/magnum-integration/pull/115 because memory usage is now low enough for this to not matter.

ImGui still supports alpha-only textures, so it might be worth revisiting. For now only RGBA is supported, this matches what the official ImGui GL backend does.

## Todo

- [x] Tests
  - [x] Check that it compiles with 1.92.9 on CI
  - [x] Can re-use the text rendering test from https://github.com/mosra/magnum-integration/pull/115 
  - [x] Fix memory corruption due to mismatching defines of `IMGUI_DISABLE_OBSOLETE_FUNCTIONS`
  - [x] Clipboard context-awareness change results in random test failure (possibly fixed with the above, can't repro anymore)
  - [x] DPI scaling isn't working as it should, add a test
  - [x] Add support for single-channel texture format

pezcode committed:
ImGuiIntegration: test text rendering.

pezcode committed:
ImGuiIntegration: add support for dynamic textures.

pezcode commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -48,12 +49,46 @@
+    const Vector2i offset = rect.min();
+    PixelStorage storage;
+    /* Data is tightly packed and rect can have any size */
+    storage.setAlignment(1);
+    storage.setRowLength(texture.Width);
+    storage.setSkip({offset.x(), offset.y(), 0});
```
Is this safe on every platform, specifically WebGL1? The ImGui code does some manual packing into a tight buffer

pezcode replied:
I copypasted what `GlyphCacheGL` does, kinda cleans up nicely when combined with only updating the combined rect

pezcode committed:
ImGuiIntegration: adapt to font-related changes in ImGui 1.92.

pezcode committed:
ImGuiIntegration: adapt to platform-related changes in ImGui 1.91.1.

pezcode committed:
ImGuiIntegration: adapt to widget changes in ImGui 1.91.9.

mosra commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -9,7 +9,7 @@
     Copyright © 2019 bowling-allie <allie.smith.epic@gmail.com>
     Copyright © 2021 Juan Pedro Bolívar Puente <raskolnikov@gnu.org>
-    Copyright © 2022, 2024 Pablo Escobar <mail@rvrs.in>
+    Copyright © 2022, 2025 Pablo Escobar <mail@rvrs.in>
```
```suggestion
    Copyright © 2022, 2024, 2025 Pablo Escobar <mail@rvrs.in>
```

Same in the other cases.

mosra commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -271,6 +325,51 @@ void Context::drawFrame() {
+                case ImTextureStatus_WantUpdates:
+                    /* There's also the combined rect of all updates in
+                       tex->UpdateRect but the official GL backend doesn't use
+                       it either */
```
:eyes: 

I'd honestly prefer the combined rect here, especially given the overhead on WebGL (where it could be doing CPU-side copies and whatnot for each such call) in my opinion makes more sense to optimize for less GL calls than less memory copied.

With Vulkan there's IIRC a way to perform a copy of multiple rectangles in a single call so there it'd be preferred, but with GL I think it's easiest to just fire off one copy and hope for it to be DMA'd in some way.

mosra commented on src/Magnum/ImGuiIntegration/Context.hpp:
```diff
@@ -498,13 +498,28 @@ namespace Implementation {
+    static void connect(Context&, Application& application) {
+        #if IMGUI_VERSION_NUM >= 19110
+        ImGuiPlatformIO& platformIO = ImGui::GetPlatformIO();
+        platformIO.Platform_GetClipboardTextFn = [](ImGuiContext*) {
+            void* state = ImGui::GetPlatformIO().Platform_ClipboardUserData;
```
Can't you access the PlatformIO from the passed context pointer in some way?

I don't like how implicit the context is here, and looking closer I didn't have it correct originally either. It's called as `_imgui.connectApplicationClipboard(*this)` and one could originally assume that it deals with the context associated with `_imgui`, but it doesn't hold and it instead deals with whatever is "current". So, like, if I have `_imgui` and `_imgui2` in some test app, and call `connectApplicationClipboard()` on both, it'll do it just for one of the two and not both, or potentially some other context entirely, or nothing at all (crash?) if no ImGui context is current.

mosra commented on src/Magnum/ImGuiIntegration/Widgets.h:
```diff
@@ -72,18 +72,40 @@ inline ImTextureID textureId(GL::Texture2D& texture) {
+@brief Image widget displaying a @ref GL::Texture2D
+@m_deprecated_since_latest Using explicit tint and border color is no longer
+    possible in newer ImGui. Use @ref image(GL::Texture2D&, const Vector2&, const Range2D&)
+    instead.
```
Can you mention this in `doc/changelog-integration.dox`, potentially updating other deprecation notices for `image()`?

And isn't the same case with tint / border in `imageButton()`?

pezcode replied:
Added. `ImGui::ImageButton()` still has the explicit tint and border color. I don't know why and I don't really want to know, either 🐱

pezcode replied:
To be honest, I think all of these widget wrappers should be deprecated, seeing how much pain they've already caused. And with `textureId()` now being a thing, there's really no good reason to use these.

mosra reviewed (commented):
Regarding GLES test failures --- ah crap, I'm still relying on the shitty SwiftShader for the GL tests here, better not trust that pile of poo with anything. I'll look into updating the CI to use llvmpipe like elsewhere.

And yes I can use different imgui versions on different CI jobs, there's plenty of them. For example the GLES3 / non-deprecated builds could use the new versions, and desktop + GLES2 the oldest supported one. Don't want to mix it up too much tho, as with more variations it'd become hard to tell what exactly is failing, whether it's due to the OS, to the platform, to ImGui version or due to something else entirely.

pezcode committed:
ImGuiIntegration: update texture with combined update rect.

pezcode committed:
ImGuiIntegration: update entire texture on WebGL1/GLES2 without EXT_unpack_subimage.

pezcode committed:
ImGuiIntegration: use ImGui context of Context for handling clipboard.

pezcode committed:
ImGuiIntegration: make text rendering test work on all versions of ImGui.

pezcode committed:
doc: update changelog.

pezcode committed:
ImGuiIntegration: test a recent version of ImGui in one job.

pezcode commented:
CircleCI seems to have issues 🌵 Some failure during `apt install` but I didn't touch any of that

pezcode committed:
ImGuiIntegration: use the existing macro here.

pezcode committed:
ImGuiIntegration: adjust text rendering test thresholds.

pezcode commented:
I realized this is missing a change - the fonts are not rasterized with the DPI scaling. Don't merge yet please 🙇

mosra commented:
When you get back to this, be sure to rebase against current `master`, I switched from swiftshader to llvmpipe for better GL test coverage and adapted to various changes in the Windows CI runners that made the builds fail.

jvannugteren commented:
Hi pezcode, any news on this? I'm actually getting a segfault when when the font texture gets updated on a relayout. This happens for example when I turn the monitor off and on again. I assume this is related to "the fonts are not rasterized with the DPI scaling".

pezcode commented:
> Hi pezcode, any news on this? I'm actually getting a segfault when when the font texture gets updated on a relayout. This happens for example when I turn the monitor off and on again. I assume this is related to "the fonts are not rasterized with the DPI scaling".

Sorry, I kind of forgot about this PR. I'll try to wrap this up.

Re your segfault: I also occasionally saw this with the old font atlas but that was DPI scale somewhere along the way becoming 0. But I'll try your repro 🙇

jvannugteren commented:
Thanks for having a look. Some further information. Stacktrace reveals that the segfault occurs on `SetSubImage` on line 212 in the `relayout` function in MagnumIntegration/src/magnum/ImguiIntegration/Context.cpp (both with SDL2 and SDL2-compat). I wonder whether the font texture even needs updating anymore, as imgui seems to take care of it by itself nowadays, which might be the problem as magnum is interfering. Also the functions from imgui `ClearTexData` is now marked [OBSOLETE] and `SetTexID` is marked legacy.

pezcode committed:
ImGuiIntegration: adapt to draw list callback changes in ImGui 1.92.8.

pezcode committed:
ImGuiIntegration: adapt to draw list count changes in ImGui 1.92.9.

pezcode committed:
ImGuiIntegration: test with IMGUI_DISABLE_OBSOLETE_FUNCTIONS set globally.

Otherwise there might be a mismatch with the definitions in
ImGui::Sources, causing different struct sizes and stack corruption.
Didn't want to do this automatically for the whole library, some users
might get surprised by that change. So just explicitly doing it for the
non-deprecated CI test. The downside is that this requires the most
recent known/supported version to be updated periodically.

Testing the latest ImGui version in this job also means we can test the
default version in the other job again.

pezcode committed:
ImGuiIntegration: verify ABI compatibility.

All built-in ImGui backends now call this by default, might as well. We
are a backend after all. However, this won't detect all cases since it
runs in ImGuiIntegration itself. To really run this in user code (where
extra defines might change ABI), we would need to move the constructor
to Context.h. Not really worth it.

pezcode committed:
ImGuiIntegration: use existing IMGUI_HAS_TEXTURES define

Not doing this in the test because the differences are caused by font
handling changes in the same version, but not dynamic textures directly.

And simplify the reuse of platformIO, the second block implies the
first block.

pezcode committed:
ImGuiIntegration: use sizes provided in ImDrawData.

Instead of querying from ImGuiIO. Future-proofing for potential
multi-window DPI-scaling, although for single-viewport applications
these are just copied and hence identical.

pezcode committed:
ImGuiIntegration: document that atlasTexture() does nothing on 1.92

This should probably be deprecated in the future, once versions below
1.92 are not supported anymore

pezcode committed:
ImGuiIntegration: mention vector-based default font in docs.

pezcode committed:
ImGuiIntegration: fix dynamic font rasterization on 1.92 and up.

We need to set DisplayFramebufferScale for ImGui to automatically set
the rasterizer density.

It's safe to set DisplayFramebufferScale on older versions as well since
it does nothing there. It was only ever meant as a multiplier for
backends to use for calculating framebuffer size from display size.

Notably, the code in drawFrame() was incorrect because it scaled twice
if FramebufferScale was not 1, once using ScaleClipRects() and then
manually using scaled(_supersamplingRatio).

pezcode committed:
ImGuiIntegration: document that manual font scaling is only needed before ImGui 1.92.

pezcode committed:
ImGuiIntegration: test HiDPI font scaling.

pezcode commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -48,12 +49,91 @@
+        .setMinificationFilter(SamplerFilter::Linear)
+        .setMagnificationFilter(SamplerFilter::Linear)
+        .setWrapping(GL::SamplerWrapping::ClampToEdge)
+        .setStorage(1, GL::TextureFormat::RGBA8, size);
```
Todo: Needs special handling for GLES2, see old code

pezcode commented on src/Magnum/ImGuiIntegration/Test/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@
 #   Copyright © 2018 Jonathan Hale <squareys@googlemail.com>
-#   Copyright © 2022 Pablo Escobar <mail@rvrs.in>
+#   Copyright © 2022, 2025 Pablo Escobar <mail@rvrs.in>
 #
```
Todo: Bump these to 2026

pezcode commented on src/Magnum/ImGuiIntegration/Test/CMakeLists.txt:
```diff
@@ -63,6 +63,7 @@ if(MAGNUM_BUILD_GL_TESTS)
             ContextTestFiles/draw.png
             ContextTestFiles/draw-scissor.png
             ContextTestFiles/draw-texture.png
+            ContextTestFiles/draw-text.png
```
Todo: Add `draw-text-dpi-scaled.png` here as well

pezcode replied:
This is causing the [ci/circleci: android-x86-gles2](https://circleci.com/gh/mosra/magnum-integration/3592) failure. Not sure why the Emscripten tests passed, investigate.

mosra replied:
This is because the Emscripten tests are not run inside a browser on the CI (too much headache involved). If you'd run them locally in a browser, it'd fail as well.

pezcode replied:
Ah oke. Fixed this issue, but I'll make sure to run it locally as well.

pezcode committed:
ImGuiIntegration: add support for single-channel textures on GL and GLES 3.0.

pezcode committed:
ImGuiIntegration: bump copyright years.

pezcode committed:
ImGuiIntegration: also test a version with dynamic textures on GLES2.

pezcode commented on src/Magnum/ImGuiIntegration/Context.h:
```diff
@@ -452,6 +458,11 @@ class MAGNUM_IMGUIINTEGRATION_EXPORT Context {
         /**
          * @brief Font texture used in `ImFontAtlas`
          * @m_since_{integration,2020,06}
+         *
+         * Returns the underlying texture for the ImGui font atlas. ImGui
+         * versions 1.92 and later create and destroy textures dynamically, on
+         * those versions this function returns an empty, default-constructed
+         * texture.
          */
         GL::Texture2D& atlasTexture() { return _texture; }
```
Should probably be deprecated once versions before 1.92 are not supported anymore

mosra commented on package/ci/circleci.yml:
```diff
@@ -636,7 +649,10 @@ jobs:
+    # Test newer ImGui in at least one GLES2 job to cover the dynamic texture
+    # path in 1.92 and up
+    - install-imgui:
+        version: "v1.92.9"
```
I'd do this in the `linux-gles2` job instead, the Android jobs are running with a weird SwiftShader GL emulation that has a ton of bugs and it's better to have that code path tested on something reliable instead.

mosra commented on src/Magnum/ImGuiIntegration/Test/ContextGLTest.cpp:
```diff
@@ -26,21 +26,13 @@
-#define IMGUI_DISABLE_OBSOLETE_FUNCTIONS
-#endif
-
+#include <cstring>
```
```suggestion
#include <cstring> /* std::strcpy() */
```

mosra commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -48,12 +49,117 @@
+
+void destroyTexture(ImTextureData& texture) {
+    CORRADE_UNUSED GL::Texture2D glTexture = GL::Texture2D::wrap(GLuint(texture.GetTexID()),
+        GL::ObjectFlag::Created|GL::ObjectFlag::DeleteOnDestruction);
```
Instead of assigning to an unused variable and the UNUSED annotation, I'd call `wrap()` alone, ignore the return value, and add a comment about what is it doing.

mosra commented on src/Magnum/ImGuiIntegration/Context.cpp:
```diff
@@ -48,12 +49,117 @@
+    storage.setAlignment(1);
+    #ifdef MAGNUM_TARGET_GLES2
+    #ifndef MAGNUM_TARGET_WEBGL
+    if(!GL::Context::current().isExtensionSupported<GL::Extensions::EXT::unpack_subimage>())
```
In order to explicitly test this code path on the CI, you could do something similar to what I did for Magnum in mosra/magnum@d5df35599f0a65e94b5745ca4a64438305166db7. `GL_EXT_texture_rg` isn't being used here for anything I think, so it'd be just one addition to the relevant files in `package/`.

mosra commented on src/Magnum/ImGuiIntegration/Context.h:
```diff
@@ -452,6 +458,11 @@ class MAGNUM_IMGUIINTEGRATION_EXPORT Context {
+         * those versions this function returns an empty, default-constructed
+         * texture.
          */
         GL::Texture2D& atlasTexture() { return _texture; }
```
I wonder if this should be a `NoCreate`d instance instead, or maybe completely compiled out on 1.92+? It's an extra GL object that isn't used for anything by ImGui anymore, and if the user would attempt to read it or add anything to it, it silently wouldn't do anything, which is I think worse than getting an error at compile time that this isn't useful for anything anymore.

Hmm, or maybe keep it but mark as deprecated if building with ImGui 1.92+, and remove altogether for 1.92+ on non-deprecated builds?

pezcode replied:
Actually it already is `NoCreate`d, the comment is incorrect 🙊  Your last suggestion probably makes the most sense, will do that.

pezcode replied:
I implemented this, but the slightly ugly side product is that Context.h now needs to include imgui.h to check for the version (or more precisely, the `IMGUI_HAS_TEXTURES` macro) on non-deprecated versions.

mosra reviewed (commented):
Will go through this more thoroughly later, this is just the immediate things that I noticed. Thanks a lot for this!

mosra commented on src/Magnum/ImGuiIntegration/Test/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@
 #             Vladimír Vondruš <mosra@centrum.cz>
 #   Copyright © 2018 Jonathan Hale <squareys@googlemail.com>
-#   Copyright © 2022 Pablo Escobar <mail@rvrs.in>
+#   Copyright © 2022, 2026 Pablo Escobar <mail@rvrs.in>
```
Could keep 2025 in all these occurences as well, in addition to 2026 :)

That's the only remaining comment I have, went commit-by-commit with a fresh brain and everything looks great, first class work once again. Thank you.

mosra commented:
Thank you! :)
